libyul/optimiser/Semantics.cpp

```diff
@@ -83,6 +88,38 @@ void MovableChecker::visit(Statement const&)
 	assertThrow(false, OptimizerException, "Movability for statement requested.");
 }
 
+
+bool InvalidationChecker::invalidatesStorage(Dialect const& _dialect, Block const& _block)
+{
+	InvalidationChecker ic(_dialect);
+	ic(_block);
+	return ic.m_invalidates;
+}
+
+bool InvalidationChecker::invalidatesStorage(Dialect const& _dialect, Expression const& _expression)
+{
+	InvalidationChecker ic(_dialect);
+	ic.visit(_expression);
+	return ic.m_invalidates;
+}
+
+void InvalidationChecker::operator()(FunctionalInstruction const& _fun)
+{
+	if (eth::SemanticInformation::invalidatesStorage(_fun.instruction))
+		m_invalidates = true;
+}
+
+void InvalidationChecker::operator()(FunctionCall const& _functionCall)
+{
+	if (BuiltinFunction const* f = m_dialect.builtin(_functionCall.functionName.name))
+	{
+		if (f->invalidatesStorage)
+			m_invalidates = true;
+	}
+	else
+		m_invalidates = true;
```

Comment:
Both true?

Reply:
Yes, the first one is a built-in function that we know invalidates storage, the second is a user-defined function where we don't know what it does and assume it invalidates storage. At some point, we should determine these properties for user-defined functions.